### ISSUE TITLE
false positive with `Self` in `Enum`

### ISSUE THREAD

**Issue by KotlinIsland** [DetachHead/basedpyright#1768]
### Description

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&reportUnannotatedClassAttribute=false&code=GYJw9gtgBApgdgV2gSwgBzCALlAooiAKFEiiwE81k4BzKVDbKAZRgBthDCBjNgQwDOAqABUYArPiQAKKRACUALkJRVUAIJQAvFAAMKtQCFtUAIwHVAYRMAmC1HsABXoIEQYWABZgAJvZ8wwFA0HgD6vj7SvALyUAC0AHxQbMgSANqsHAC6ymp5UCAeCCBwUGnRAHSGWYRAA)

```python
from enum import Enum
from typing import Self

class TestEnum(Enum):
    A = 0
    B = 1
    C = 2
    
    @classmethod
    def get_odd(cls) -> list[Self]:
        return [cls.B]

```

`Enum` with members are final, so `Self` should account for this

**Comment by DetachHead** [comment]
not just enums, anything decorated with `@final`

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&reportUnannotatedClassAttribute=false&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbAA0UwqAhiQFD0AC1Kd9AxiTQM49QAxMGABc9KBKgATIsCooAFD1LAAlFAC0APkIqxkg1BBEYAVxApBwhaqA)

```python
from typing import Self, final

@final
class Foo:
    def fn(self) -> Self:
        return Foo() # error: Type "Foo" is not assignable to return type "Self@Foo"
```
